File: php/lib/functional_list.php
<?php

abstract class FunctionalList {
    public static function nil() {
        return new Nil();
    }

    public static function cons($head, $tail) {
        return new Cons($head, $tail);
    }

    public static function of() {
        $args = func_get_args();
        return static::_from_array($args, 0, count($args));
    }

    public static function from_array($items)
    {
        return self::_from_array($items, 0, count($items));
    }

    private static function _from_array($items, $start, $end) {
        if ($start === $end) {
            return static::nil();
        }
        return static::cons($items[$start], static::_from_array($items, $start + 1, $end));
    }

    public abstract function isEmpty();
    public abstract function head();
    public abstract function tail();

    public abstract function map($f);
}

class Nil extends FunctionalList {
    public function isEmpty() {
        return true;
    }

    public function head() {
        throw new Exception('Nil does not have a head');
    }

    public function tail() {
        throw new Exception('Nil does not have a tail');
    }

    public function map($f) {
        return $this;
    }
}

class Cons extends FunctionalList {
    public function __construct($head, $tail) {
        $this->head = $head;
        $this->tail = $tail;
    }

    public function isEmpty() {
        return false;
    }

    public function head() {
        return $this->head;
    }

    public function tail() {
        return $this->tail;
    }

    public function map($f) {
        return FunctionalList::cons($f($this->head), $this->tail->map($f));
    }
}

?>

File: python/lib/functional_list.py
class FunctionalList:
    @staticmethod
    def of(*items):
        return FunctionalList._fromArray(items, 0, len(items))

    @staticmethod
    def _fromArray(items, start, end):
        return Nil() if start == end else Cons(items[start], FunctionalList._fromArray(items, start + 1, end))

class Nil(FunctionalList):
    def isEmpty(self):
        return True

    def map(self, mapping):
        return self

    def __eq__(self, other):
        return other.isEmpty()

class Cons(FunctionalList):
    def __init__(self, head, tail):
        self._head = head
        self._tail = tail

    def isEmpty(self):
        return False

    @property
    def head(self):
        return self._head

    @property
    def tail(self):
        return self._tail

    def map(self, mapping):
        return Cons(mapping(self.head), self.tail.map(mapping))

    def __eq__(self, other):
        return not other.isEmpty() \
            and self.head == other.head \
            and self.tail == other.tail
